### PR TITLE
Cranelift: implement `heap_{load,store}` instruction legalization

### PR DESCRIPTION
**Pull request by fitzgen**
This does not remove `heap_addr` yet, but it does factor out the common
bounds-check-and-compute-the-native-address functionality that is shared between
all of `heap_{addr,load,store}`.

Finally, this adds a missing optimization for when we can dedupe explicit bounds
checks for static memories and Spectre mitigations.

### CODE REVIEW

**Comment by fitzgen** [comment]
Okay I think this can land as-is and I can remove `uload8` et al in a follow up (as discussed in the Cranelift meeting today).

@cfallin mind taking a look at this?

Will post sightglass benchmarks in a minute.

File: cranelift/codegen/src/legalizer/heap.rs [hunk 1]
```diff
@@ -6,268 +6,394 @@
...
-            spectre_oob_comparison = Some(SpectreOobComparison {
-                cc: IntCC::UnsignedGreaterThan,
-                lhs: index,
-                rhs: limit,
-            });
+    // Finally, the following right-hand sides of the matches do have a little
+    // bit of duplicated code across them, but I think writing it this way is
+    // worth it for readability and seeing very clearly each of our cases for
+    // different bounds checks and optimizations of those bounds checks. It is
+    // intentinally written in a straightforward case-matching style that will
```
Comment (cfallin) [inline cranelift/codegen/src/legalizer/heap.rs:147]: s/intentinally/intentionally/

**Comment by fitzgen** [comment]
Sightglass reports no differences.

<details>

```
execution :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  No difference in performance.

  [8508043 11057644.93 37220741] main.so
  [8724611 9739249.27 18570033] with-guard-pages.so

instantiation :: cycles :: benchmarks/bz2/benchmark.wasm

  No difference in performance.

  [141659 246194.93 432381] main.so
  [143198 272550.67 604659] with-guard-pages.so

instantiation :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [465184 554304.70 1001114] main.so
  [467711 586950.90 903247] with-guard-pages.so

instantiation :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  No difference in performance.

  [178842 284227.00 689740] main.so
  [183614 271998.40 419997] with-guard-pages.so

execution :: cycles :: benchmarks/bz2/benchmark.wasm

  No difference in performance.

  [118533772 128097726.07 191789854] main.so
  [117856595 125177352.07 178670691] with-guard-pages.so

compilation :: cycles :: benchmarks/bz2/benchmark.wasm

  No difference in performance.

  [189359133 214237253.40 279159745] main.so
  [192812151 217913618.83 267477996] with-guard-pages.so

compilation :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  No difference in performance.

  [282374972 334814772.03 518486262] main.so
  [297502265 337653688.23 556526555] with-guard-pages.so

execution :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [1045855548 1072070714.57 1143533825] main.so
  [1047025362 1069661379.83 1112794276] with-guard-pages.so

compilation :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [6898333529 7047898225.47 7214123481] main.so
  [6890104965 7059359428.33 7328196887] with-guard-pages.so
```

</details>

**Review by cfallin — APPROVED** [review]
This all looks great to me. I just have one typo fix; no issues with the logic that I can see.

I really appreciate the case breakdown -- thank you for this very clear refactor!

**Pull request merged**